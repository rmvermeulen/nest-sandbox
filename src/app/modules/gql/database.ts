type ID = string
export interface Author {
  id: ID
  firstName: string
  lastName: string
  postIds: Array<Post['id']>
}

export interface Post {
  id: ID
  title: string
  authorId: Author['id']
  votes: number
}

export const posts: Post[] = [
  { id: 'post-1', title: 'cool stuff', authorId: 'author-1', votes: 12 },
  { id: 'post-2', title: 'cooler stuff', authorId: 'author-2', votes: 2 },
  { id: 'post-3', title: 'coolest stuff', authorId: 'author-1', votes: 5 },
]

export const authors: Author[] = [
  {
    id: 'author-1',
    firstName: 'Andy',
    lastName: 'Armstrong',
    postIds: ['post-1', 'post-3'],
  },
  {
    id: 'author-2',
    firstName: 'Bob',
    lastName: 'Barley',
    postIds: ['post-2'],
  },
]
