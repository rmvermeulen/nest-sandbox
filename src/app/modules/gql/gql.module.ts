import {
  MiddlewaresConsumer,
  Module,
  NestModule,
  RequestMethod,
} from '@nestjs/common'
import { GraphQLFactory, GraphQLModule } from '@nestjs/graphql'
import { graphiqlExpress, graphqlExpress } from 'apollo-server-express'
import { Request as Req, Response as Res } from 'express'

@Module({
  imports: [GraphQLModule],
})
export class GQLModule implements NestModule {
  constructor(private readonly graphQLFactory: GraphQLFactory) {}

  public configure(consumer: MiddlewaresConsumer) {
    const typeDefs = this.graphQLFactory.mergeTypesByPaths('./**/*.graphql')
    const schema = this.graphQLFactory.createSchema({ typeDefs })

    consumer
      .apply(
        graphqlExpress(req => ({
          schema,
          rootValue: req,
        })),
      )
      .forRoutes({ method: RequestMethod.ALL, path: '/graphql' })
      .apply(graphiqlExpress(() => ({ endpointURL: '/graphql' })))
      .forRoutes({
        method: RequestMethod.ALL,
        path: '/graphiql',
      })
      // route everything else to the IDE
      .apply((req: Req, res: Res) => res.redirect('/graphiql'))
      .forRoutes({ method: RequestMethod.ALL, path: '*' })
  }
}
