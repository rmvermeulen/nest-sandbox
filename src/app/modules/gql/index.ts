export { GQLModule } from './gql.module'
