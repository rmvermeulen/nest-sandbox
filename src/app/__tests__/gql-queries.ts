import { Test } from '@nestjs/testing'
import * as express from 'express'
import { path } from 'ramda'
import * as request from 'supertest'

import { GQLModule } from '@modules/gql'

describe('Server', () => {
  let server: Express.Application

  beforeAll(async () => {
    const testModule = await Test.createTestingModule({
      imports: [GQLModule],
    })
      // .overrideComponent(database)
      // .useValue({})
      .compile()

    server = express()
    const app = testModule.createNestApplication(server)
    await app.init()
  })

  it(`serves the IDE`, () =>
    request(server)
      .get('/graphiql')
      .expect(200))

  // graphql helper
  const getQuery = (query: string) =>
    request(server)
      .post('/graphql')
      .send({ query })
      .expect(200)
      .then(path(['body', 'data']))

  it('can query Posts', () =>
    expect(
      getQuery(`{
      posts { id }
    }`),
    ).resolves.toMatchSnapshot())

  it('can query Authors', () =>
    expect(
      getQuery(`{
      authors { id }
    }`),
    ).resolves.toMatchSnapshot())

  it('can resolve arbitrarily nested queries', () =>
    expect(
      getQuery(`{
        authors {
          posts {
            author {
              posts {
                author {
                  id
                }
              }
            }
          }
        }
      }`),
    ).resolves.toMatchSnapshot())
})
