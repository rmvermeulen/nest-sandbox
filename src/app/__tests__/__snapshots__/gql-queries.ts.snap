// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Server can query Authors 1`] = `
Object {
  "authors": null,
}
`;

exports[`Server can query Posts 1`] = `
Object {
  "posts": null,
}
`;

exports[`Server can resolve arbitrarily nested queries 1`] = `
Object {
  "authors": null,
}
`;
